UefiDbgExt/modules.cpp

```diff
@@ -199,56 +229,109 @@ findall (
   )
 {
   HRESULT  Result;
-  ULONG64  BsPtrAddr;
-  ULONG64  BsTableAddr;
+  ULONG64  SystemPtrAddr;
+  ULONG64  SystemTableAddr;
+  ULONG64  Signature            = 0;
+  ULONG    BytesRead            = 0;
+  ULONG64  SystemTableSignature = (('I') | (static_cast<ULONG64>('B') << 8) | (static_cast<ULONG64>('I') << 16) | (static_cast<ULONG64>(' ') << 24) | (static_cast<ULONG64>('S') << 32) | (static_cast<ULONG64>('Y') << 40) | (static_cast<ULONG64>('S') << 48) | (static_cast<ULONG64>('T') << 56));
+  PSTR     Response;
 
   INIT_API ();
 
-  if (gUefiEnv != DXE) {
-    dprintf ("Only supported for DXE!\n");
+  if ((gUefiEnv != DXE) && (gUefiEnv != RUST)) {
+    dprintf ("Only supported for DXE and Rust!\n");
     return ERROR_NOT_SUPPORTED;
   }
 
   //
-  // First find the current module
+  // First find the current module. We only do this to see if we are in the core to find the system table pointer
+  // symbols. If we are not in the core, we will ask the monitor for the system table pointer address, failing that
+  // we will scan memory for the EFI_SYSTEM_TABLE_SIGNATURE, per UEFI spec. The C DXE environment does not have the
+  // monitor command, so relies on the core symbols having been loaded at the initial breakpoint (or us being broken
+  // into the core now)
   //
 
-  Result = FindModuleBackwards (GetExpression ("@$ip"));
-  if (Result != S_OK) {
-    return Result;
-  }
+  FindModuleBackwards (GetExpression ("@$ip"));
+
+  g_ExtControl->Execute (
+                  DEBUG_OUTCTL_ALL_CLIENTS,
+                  "ld *ore*",
+                  DEBUG_EXECUTE_DEFAULT
+                  );
 
   //
-  // Find the core module. This might be the same as the executing one.
+  // Find the system table pointer, which we may not find if we are not in the core module
   //
 
-  BsPtrAddr = GetExpression ("gBS");
-  if (BsPtrAddr == NULL) {
-    dprintf ("Failed to find boot services table pointer!\n");
-    return ERROR_NOT_FOUND;
+  if (gUefiEnv == DXE) {
+    SystemPtrAddr = GetExpression ("mDebugTable");
+    if (!ReadPointer (SystemPtrAddr, &SystemPtrAddr)) {
+      dprintf ("Failed to read memory at %llx to get system table from ptr\n", SystemPtrAddr);
+      return ERROR_NOT_FOUND;
+    }
+  } else if (gUefiEnv == RUST) {

```

Comment:
Should you account for the scenario where this is rust but not the debugger? e.g. QEMU debugging

Not a blocking ask, but worth thinking about